### ISSUE TITLE
Backend API allows archival_objects to be created with no resource ID

### ISSUE THREAD

**Issue by mistydemeo** [archivesspace/archivesspace#265]
The backend API allows new archival objects to be created which have no specified resource ID. If the new archival object is the child of an existing archival object, it's not filled in using that; it remains null.

While the backend API accepts these records, the frontend isn't designed to be able to deal with a record with no resource ID, and trying to view any of these records results in a 500 with the following backtrace:

```
NoMethodError (undefined method `[]' for nil:NilClass):
  app/models/resolver.rb:47:in `default_uri_properties'
  app/models/resolver.rb:25:in `view_uri'
  app/controllers/resolver_controller.rb:25:in `resolve_readonly'
```

The line here is unconditionally referencing the archival object's resource: https://github.com/archivesspace/archivesspace/blob/master/frontend/app/models/resolver.rb#L47

Which is a reasonable assumption - maybe the backend should just reject a record like this?

POSTing the following object reproduces the problem:

``` json
{
  "level": "series",
  "parent": {
    "ref": "/repositories/2/archival_objects/2"
  },
  "jsonmodel_type": "archival_object",
  "title": "Test no resource"
}
```

**Comment by cfitz** [comment]
Hi,

Sorry about the delay. I made the mistake of thinking I could take a vacation and I've been paying for it since...

So, yeah, this is not good. Looking at this =>
https://github.com/archivesspace/archivesspace/blob/master/backend/app/model/mixins/tree_nodes.rb#L257-L289
I think what is supposed to happen is that the TreeNode will be added as an "orphan" is the root_record_type ( in the case of AO, it's a resource ) is not included. However, this looks like it doesn't handle the use case your describing, where the parent is included ( so the "orphan" parent_name doesn't get applied to the incoming JSON ). 

So..trying to think what the best use case is. We could:
1) Reject these AOs from being created ( require the Resource )
2) Accept them, but make them orphans
3) Try and pull the resource from the parent and apply it. 

Maybe #3 is the best considering #264 . We could apply it and validate that the parent's resource_id matches the child. 

Any Thoughts, 
@bdwestbrook  @quoideneuf @marktriggs @jambun ?

**Comment by mistydemeo** [comment]
I think 3) sounds like the best option; it fits best with what an API user would expect, and I don't think there are any ambiguities it introduces.

**Comment by quoideneuf** [comment]
What if the incoming archival object has a root record id that is different from its parent's? It seems like #1 might be more straightforward - just don't allow incoming archival objects that have a parent unless they also have the same resource as the parent.

**Comment by lmcglohon** [comment]
This got fixed with 2.0.0-RC1.

**Comment by westoneb** [comment]
what if you already have orphaned records from before the 2.0.0 fix? We appear to be experiencing this problem. We have 210 archival_object records that have neither a parent_id or root_record_id. AS 1.5.3 apparently ignores these (they show up in a search, but cannot be opened) but when we upgraded a test instance to 2.0.0 and later 2.1.0, the indexer fails when it gets to one of these corrupted records, and starts over. This causes AS to be in a perpetual indexing loop, and the CPU load on the server stays very high. The records don't appear in the frontend interface, and I could not delete them with the API, which throws the same "key not found" error that the indexer throws.
